spec/graphql/execution/breadth_runtime_spec.rb

```diff
@@ -0,0 +1,325 @@
+# frozen_string_literal: true
+require "spec_helper"
+
+describe "GraphQL::Execution::Interpreter for breadth-first execution" do
+  # A breadth-first interpreter uses the following runtime interface:
+  # - evaluate_selection(result_key, ast_nodes, selections_result)
+  # - exit_with_inner_result?
+  class SimpleBreadthRuntime < GraphQL::Execution::Interpreter::Runtime
+    class BreadthObject < GraphQL::Execution::Interpreter::Runtime::GraphQLResultHash
+      attr_accessor :breadth_index
+    end
+
+    def initialize(query:)
+      query.multiplex = GraphQL::Execution::Multiplex.new(
+        schema: query.schema,
+        queries: [query],
+        context: query.context,
+        max_complexity: nil,
+      )
+
+      super(query: query, lazies_at_depth: Hash.new { |h, k| h[k] = [] })
+      @breadth_results_by_key = {}
+    end
+
+    def run
+      result = nil
+      query.current_trace.execute_multiplex(multiplex: query.multiplex) do
+        query.current_trace.execute_query(query: query) do
+          result = yield
+        end
+      end
+      result
+    ensure
+      delete_all_interpreter_context
+    end
+
+    def evaluate_breadth_selection(objects, parent_type, node)
+      result_key = node.alias || node.name
+      @breadth_results_by_key[result_key] = Array.new(objects.size)
+      objects.each_with_index do |object, index|
+        app_value = parent_type.wrap(object, query.context)
+        breadth_object = BreadthObject.new(nil, parent_type, app_value, nil, false, node.selections, false, node, nil, nil)
+        breadth_object.ordered_result_keys = []
+        breadth_object.breadth_index = index
+
+        state = get_current_runtime_state
+        state.current_result_name = nil
+        state.current_result = breadth_object
+        @dataloader.append_job { evaluate_selection(result_key, node, breadth_object) }
+      end
+
+      @dataloader.run
+      GraphQL::Execution::Interpreter::Resolve.resolve_each_depth(@lazies_at_depth, @dataloader)
```

Comment:
In the actual prototype, I’m running dataloader inline with the field but this lazy step runs out-of-band after a bunch of fields have been queued. Sounds like with https://github.com/rmosolgo/graphql-ruby/pull/5422, I’d just run dataloader once in that out-of-band position?